### ISSUE TITLE
Support Generic parameters

### ISSUE THREAD

**Issue by nezda** [immutables/immutables#282]
Making an actual issue from http://immutables.github.io/immutable.html#generics-not-supported which says

> This is definitely a limitation and probably will be lifted at some point.

This would allow us to avoid mixed usage of `@Value.Immutable` and `@AutoValue` in our code, as they support this: https://github.com/google/auto/tree/master/value#generics.  Maybe this project could borrow from their implementation?

Classic example:

```java
  @Value.Immutable
  public abstract static class NamedTuple<F, B> {
    public abstract F foo();
    public abstract B bar();
  }
```

**Comment by elucash** [comment]
Maybe it's time to introduce generics. It still may not work with other stuff (like Gson TypeAdapter generator) from the beginning, but that may be fixed later. Thank you for the feature request!

**Comment by elucash** [comment]
2.2-beta released with generics support, inviting to evaluate it

**Comment by nezda** [comment]
Awesome @elucash!

**Comment by nezda** [comment]
I tried out 2.2-beta-5 and this feature seems to work as I'd expect in the sense that it compiles and generates expected factories, etc., however I did get a some new test failures in my app caused by #326.  Thanks again for this awesome library!

**Comment by elucash** [comment]
released in 2.2 Thank you!
